File: module_1/python/day_4/is_bipartite.py
"""
Title: Determine if a given graph is bipartite.

Problem:
    Given an undirected graph, return true if and only if it is bipartite.

    Recall that a graph is bipartite if we can split it's set of nodes into two
    independent subsets A and B such that every edge in the graph has one node in A
    and another node in B.

    The graph is given in the following form: graph[i] is a list of indexes j for
    which the edge between nodes i and j exists.  Each node is an integer between 0
    and graph.length - 1.  There are no self edges or parallel edges: graph[i] does
    not contain i, and it doesn't contain any element twice.

Execution: python is_bipartite.py
"""
import unittest


def is_bipartite(graph: list):
    """Function for checking if graph is bipartite."""
    color = dict()
    def depth_first_search(position: int):
        """Use depth first search to determine if graph is bipartite."""
        for i in graph[position]:
            if i in color:
                if color[i] == color[position]:
                    return False
            else:
                color[i] = 1 - color[position]
                if not depth_first_search(i):
                    return False
        return True
    for i in range(len(graph)):
        if i not in color:
            color[i] = 0
            if not depth_first_search(i):
                return False
    return True


class TestIsBipartite(unittest.TestCase):
    """Unit test for is_bipartite."""

    def test_1(self):
        """
        Test for graph:
            0----1
            |    |
            |    |
            3----2
        """
        graph = [[1, 3], [0, 2], [1, 3], [0, 2]]
        self.assertEqual(is_bipartite(graph), True)

        output = """
        The graph looks like this:
            0----1
            |    |
            |    |
            3----2
        We can divide the vertices into two groups: {0, 2} and {1, 3}.
        """
        print(f"Explanation: {output}")

    def test_2(self):
        """
        Test for graph:
            0----1
            | \  |
            |  \ |
            3----2
        """
        graph = [[1, 2, 3], [0, 2], [0, 1, 3], [0, 2]]
        self.assertEqual(is_bipartite(graph), False)
        output = """
        The graph looks like this:
            0----1
            | \  |
            |  \ |
            3----2
        We cannot find a way to divide the set of nodes into two 
        independent subsets.
        """
        print(f"Explanation: {output}")


if __name__ == '__main__':
    unittest.main()

File: module_1/java/IsBipartite.java
/*
 *   Title: Is Graph Bipartite
 *   Leetcode Link: https://leetcode.com/problems/is-graph-bipartite/
 *
 *   Problem: Given an undirected graph, return true if and only if it is
 *   bipartite.
 *
 *   Recall that a graph is bipartite if we can split it's set of nodes into two
 *   independent subsets U and V such that every edge in the graph has one node
 *   in U and another node in B.
 *
 *   The graph is given in the following form:
 *      graph[i] is a list of indexes j for which the edge between nodes i and j
 *      exists.  Each node is an integer between 0 and graph.length - 1.  There
 *      are no self edges or parallel edges: graph[i] doesnot contain i, and it
 *      doesn't contain any element twice.
 *
 *   Input:
 *      int[][] graph       => Edge list representing graph
 *   Output:
 *      boolean             => True if and only if graph is bipartite
 *
 *   Execution: javac IsBipartite.java && java -ea IsBipartite
 */

import java.util.*;

public class IsBipartite {

    /*
     * We select a random starting node and then  greedily assign each node to
     * either U or V. We do this using DFS. As we do this, we look for
     * contradictions. If we can successfully assign all the nodes, then the
     * graph is bipartite.
     *
     * Time Complexity: O(V+E)
     * Space Complexity: O(V)
     */
    public static boolean isBipartite(int[][] graph) {
        // Track the coloring of each node that we visit. For simplicity, nodes
        // in U will be set to 1 and nodes in V will be -1
        int[] colors = new int[graph.length];

        // Our nodes aren't all necessarily connected, so we need to be sure to
        // iterate over every connected component of the graph
        for (int i = 0; i < graph.length; i++) {
            if (!isBipartite(graph, i, colors)) return false;
        }

        return true;
    }

    /*
     * Private inner function
     */
    private static boolean isBipartite(int[][] graph, int curr, int[] colors) {
        // If the node hasn't been visited yet we need to initialize it. The
        // value doesn't matter
        if (colors[curr] == 0) colors[curr] = 1;

        // Use DFS to visit all nodes. Each time we visit a node, we'll set
        // it to be the opposite color. As we do this we look for
        // contradictions
        for (int next : graph[curr]) {
            // If the next node hasn't been visited, set it to the opposite of
            // the current color. Then continue DFS. Otherwise, if it has been
            // visited make sure it's the right color
            if (colors[next] == 0) {
                colors[next] = -colors[curr];
                if (!isBipartite(graph, next, colors)) return false;
            } else if (colors[next] != -colors[curr]) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        int[][] arr1 = { { 1, 3 }, { 0, 2 }, {1, 3}, {0, 2} };
        assert isBipartite(arr1) == true;

        int[][] arr2 = { {1, 2, 3}, { 0, 2 }, {0, 1, 3}, {0, 2} };
        assert isBipartite(arr2) == false;

        // ADD TEST CASES HERE

        System.out.println("Passed all test cases");
    }

}
